server/filters.go

```diff
@@ -0,0 +1,154 @@
+)
+
+func pxcFilter(versions map[string]*pbVersion.Version, apply string, current string) error {
+	if len(versions) == 0 {
+		return status.Error(codes.Internal, "no versions to filter")
+	}
+
+	keys := make([]string, 0, len(versions))
+	for k := range versions {
+		keys = append(keys, k)
+	}
+
+	sorted, err := sortedVersionsDesc(keys)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to sort versions: %v", err)
+	}
+
+	if (strings.ToLower(apply) == recommended || strings.ToLower(apply) == latest) && current == "" {
+		deleteOtherBut(sorted[0].String(), versions)
+		return nil
+	}
+
+	desired := apply //assume version number
+	if strings.ToLower(apply) == recommended || strings.ToLower(apply) == latest {
+		desired = current
+
+		c, err := semver.NewVersion(current)
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "invalid current version: %s", current)
+		}
+
+		for _, s := range sorted {
+			if s.Equal(c) || s.LessThan(c) {
+				break
+			}
+			if versions[s.String()].Status != pbVersion.Status_disabled && c.Major() == s.Major() {
+				desired = s.String()
+			}
+		}
+	}
+
+	deleteOtherBut(desired, versions)
+	if len(versions) == 0 {
+		return status.Errorf(codes.NotFound, "version %s does not exist", desired)
+	}
+
+	return nil
+}
+
+func defaultFilter(versions map[string]*pbVersion.Version, apply string) error {
+	if len(versions) == 0 {
+		return status.Error(codes.Internal, "no versions to filter")
+	}
+
+	keys := make([]string, 0, len(versions))
+	for k := range versions {
+		keys = append(keys, k)
+	}
+
+	sorted, err := sortedVersionsDesc(keys)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to sort versions: %v", err)
+	}
+
+	if apply == "" {
+		apply = sorted[0].String()
+	}
+
+	deleteOtherBut(apply, versions)
+
+	return nil
+}
+
+func depFilter(versions map[string]interface{}, productVersion string) (string, error) {
+	if len(versions) == 0 {
+		return "", nil
+	}
+
+	keys := make([]string, 0, len(versions))
+	for k := range versions {
+		keys = append(keys, k)
+	}
+
+	sorted, err := sortedVersionsDesc(keys)
+	if err != nil {
+		return "", status.Errorf(codes.Internal, "failed to sort versions: %v", err)
+	}
+
+	desired := sorted[0].String()
+	for _, s := range sorted {
+		b, err := json.Marshal(versions[s.String()])
+		if err != nil {
+			return "", status.Errorf(codes.Internal, "failed to marshal deps logic: %v", err)
+		}
+		logic := bytes.NewReader(b)
+		data := strings.NewReader(fmt.Sprintf(`{  "productVersion" : "%s" }`, productVersion))
+
+		var result bytes.Buffer
+		err = jsonlogic.Apply(logic, data, &result)
+		if err != nil {
+			return "", status.Errorf(codes.Internal, "failed to apply logic: %v", err)
+		}
+
+		if strings.TrimSuffix(result.String(), "\n") == "true" {
+			desired = s.String()
+			break
+		}
+	}
+
+	return desired, nil
+}
+
+func deleteOtherBut(v string, versions map[string]*pbVersion.Version) {
+	for k := range versions {
+		if k != v {
```

Comment:
we need to check only `Major.Minor.Patch` here, without `-suffix`.
Reason - `suffix` returned from MySQL doesn't match to `suffix` in our matrix

Reply:
https://github.com/Percona-Lab/percona-version-service/pull/1/commits/e68e54f9c2193cb81b2ca3d1356bd9cbf475e585